### PR TITLE
1.15 Item Pickup event

### PR DESCRIPTION
**Pull request by Draylar**
Extension of #260 built for the 1.15 branch. Everything from this PR is based on the discussion from the previous one. Tell me if I missed anything~

### CODE REVIEW

File: fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerPickupItemCallback.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+/**
+ * Callback for the player picking up an item entity.
+ * Is hooked in before the item is picked up.
+ *
+ * <p>Upon return:
+ * <ul><li>SUCCESS cancels further processing and picks up the item.
```
Comment (liach) [inline fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerPickupItemCallback.java:31]: note there is another `CONSUME` enum value.

**Comment by liach** [comment]
Should be more generic to cover mob item pickup, etc.

**Comment by Draylar** [comment]
> Should be more generic to cover mob item pickup, etc.

The player picking up an item is handled separately from regular entities picking up items. In the case of the player, `ItemEntity` is responsible; in the case of entities (eg. villagers), `tickMovement` iterates over nearby items, and calls `loot` on the entity to attempt to pick it up.

I don't think we should merge the two together; the latter should be a separate event like `EntityLootItemCallback`, or something. I say this because `loot` is overridden in several entities (see: `VillagerEntity`, which is probably a key target for this callback), and there would need to be some special mixins to target each case. Another example is `PandaEntity`, which:
  1. overrides `loot`, callback wouldn't be triggered
  2. only even considers picking up an item if it's food

To properly support those, you'd probably have to have special mixins for each entity (eg. only callback in panda if the item is food), which seems out of scope for this PR.

File: fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerPickupItemCallback.java (hunk 1 above)
Comment (Draylar) [reply fabric-events-interaction-v0/src/main/java/net/fabricmc/fabric/api/event/player/PlayerPickupItemCallback.java:31]: It doesn't seem like any of the current callbacks document or handle `CONSUME`. Is there an agreed-upon standard for what it should do?

**Comment by Draylar** [comment]
General follow-up: there are a few major issues with this that I discovered while refactoring, so refer to #api for further discussion. Assume this is on hold until further commits are made.

In short: I've decided to try to split the PR into two callbacks; one handles pre-pickup, and one handles post. For your information, here's a brief overview of `ItemEntity#onPlayerCollision`, which is where most of the general logic is triggered:

```java
if the world is server:
   [store associated ItemStack/Item/count]
   if the player can pickup the item && player.inventory.InsertStack(stack) succeeds:
      // note: the saved stack's quantity is now different; is either empty/remainder
      broadcast pickup information
      if our associated ItemStack is empty:
         remove it and set it's count to the original count[?]
```

Here's an overview of my 2 callback ideas:

### PlayerAttemptPickupItemCallback:
   Listeners return a `TypedActionResult<ItemStack>` and are given access to the `PlayerEntity` and `ItemStack` which will be picked up (which is not the stack on the ground, but the one that _would_ enter the player's inventory, including proper count). Returning `FAIL` prevents the stack from being picked up, and modifying the `ItemStack` before returning it changes what the player receives. 

Use cases include:
  - modifying pickups based on items currently in the inventory
  - setting NBT tags on items, such as restarting a timer
  - preventing players from picking up certain gated items

### PlayerPickupItemCallback:
   Listeners are not responsible for returning anything; this is simply a notification callback that provides the `PlayerEntity` and newly gained `ItemStack`. 

Use cases include:
  - triggering advancement or quest systems
  - server item logging

### Issue
The issue is that this method doesn't give context on _how much_ of the stack the player will pick up before the pre-callback would be called, as that occurs much deeper down the chain. Solutions that I can think of are fairly limited, but include:
  - store pickup data separately and have the hook access it when needed
  - calculate how much will be picked up inside the mixin
  - call event for the entire stack, regardless of how much the player _can_ pick up
  - nuke method logic and have it marked as a no-go zone for other mixins :)

To take a step back, these solutions rely on the fact that the "hey, can I pick up these items?" callback gives you the amount that _would_ enter the player's inventory. If we decide this shouldn't be the callbacks job, it'll be a lot easier to implement (but functionality/practical usage is lost). It would then change `PlayerAttemptPickupItemCallback` to give you a `PlayerEntity` and `ItemEntity`, and expect an `ItemEntity` in return.

The problem with the last mentioned solution, of course, is that you couldn't directly change items going into your inventory or give them tags (as it would affect remainders on the ground as well).

**Pull request closed without merging**